### PR TITLE
FileDataloader example

### PR DESCRIPTION
**Pull request by sivasankalpp**
## What this PR does / why we need it
- Add example for FileDataloader
- Specify dataloader for loading additional seed examples in the qna.yaml file 
- `data_path` is the absolute path

For example, if you have a directory structure like this:

```
data/  
│     |___ temporal_reasoning/
|     |      |_____ qna.yaml
|     |      |_____ additional_seed_examples.yaml
```

Then you can specify seed data in your `qna.yaml` in the following way:
```yaml
created_by: ""
task_description: ""
dataloader:
  type: file
  data_path: data/temporal_reasoning/additional_seed_examples.yaml
seed_examples:
  - answer: ...
    question: ...
```

## If applicable**
- [ ] this PR contains documentation
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

### CODE REVIEW

File: fms_dgt/base/task.py [hunk 1]
```diff
@@ -91,9 +92,21 @@ def __init__(
...
-                TYPE_KEY in dataloader
-            ), f"Must specify data loader type with '{TYPE_KEY}' key"
+            assert TYPE_KEY in dataloader, (
+                "Must specify dataloader type with %s key",
+                TYPE_KEY,
+            )
+            assert DATA_PATH_KEY in dataloader, (
+                "Must specify dataloader data path with %s key",
+                DATA_PATH_KEY,
+            )
```
Comment (mvcrouse) [inline fms_dgt/base/task.py:102]: I think we should keep all the dataloader-specific checking within the dataloader, e.g., DATA_PATH_KEY may not be a required field for other dataloaders

File: fms_dgt/base/task.py [hunk 2]
```diff
@@ -91,9 +92,21 @@ def __init__(
...
+            assert seed_examples == [
+                {}
+            ], "Please specify either seed_examples or dataloader, not both"
```
Comment (mvcrouse) [inline fms_dgt/base/task.py:105]: We probably don't want to check this, as a user may have a small set of seed examples on-hand and then load the rest

File: fms_dgt/base/task.py [hunk 3]
```diff
@@ -91,9 +92,21 @@ def __init__(
...
+            dir_name = "/".join(name.split("->"))
+            dataloader[DATA_PATH_KEY] = os.path.join(
+                dir_name, dataloader[DATA_PATH_KEY]
+            )
```
Comment (mvcrouse) [inline fms_dgt/base/task.py:109]: I think we should just assume the PATH is given as an absolute path, that's a simpler setting that's easy to understand and we don't know if we'll keep that "->" path modification going forward

File: fms_dgt/base/task.py (hunk 2 above)
Comment (sivasankalpp) [reply fms_dgt/base/task.py:105]: Yeah, that was my initial thought too. In that case, we must pass this "initial" set of seed examples to the dataloader to combine with the rest.

File: fms_dgt/base/task.py (hunk 1 above)
Comment (sivasankalpp) [reply fms_dgt/base/task.py:102]: Makes sense! I'll remove this check

File: fms_dgt/base/task.py (hunk 3 above)
Comment (sivasankalpp) [reply fms_dgt/base/task.py:109]: alright, absolute path it is!

File: fms_dgt/base/task.py [hunk 4]
```diff
@@ -91,9 +92,12 @@ def __init__(
...
         else:
-            assert (
-                TYPE_KEY in dataloader
-            ), f"Must specify data loader type with '{TYPE_KEY}' key"
+            assert TYPE_KEY in dataloader, (
+                "Must specify dataloader type with %s key",
+                TYPE_KEY,
+            )
+            if dataloader[TYPE_KEY] == "file":
+                dataloader["seed_examples"] = seed_examples
```
Comment (mvcrouse) [inline fms_dgt/base/task.py:100]: I think just having seed_examples assumed to be a kwarg of dataloader would suffice here, we don't want to add logic for a particular dataloader in this task class

```suggestion
assert TYPE_KEY in dataloader, (
                "Must specify dataloader type with %s key",
                TYPE_KEY,
            )
self._dataloader = get_dataloader(dataloader.pop(TYPE_KEY))(**dataloader)
```

File: fms_dgt/dataloaders/file.py [hunk 5]
```diff
@@ -11,14 +11,22 @@
...
 @register_dataloader("file")
 class FileDataloader(DefaultDataloader):
-    """Class for all json datasets"""
+    """Class for all json/yaml datasets"""
 
-    def __init__(self, data_path: str = None) -> None:
+    def __init__(
+        self,
+        data_path: str = None,
+        seed_examples: List[Any] = [],
```
Comment (mvcrouse) [inline fms_dgt/dataloaders/file.py:24]: Usually want to initialize with `None` as a default arg, then set the value in the code itself (https://nikos7am.com/posts/mutable-default-arguments/)

File: fms_dgt/base/task.py (hunk 4 above)
Comment (sivasankalpp) [reply fms_dgt/base/task.py:100]: I'm a little confused. I get that we don't want a dataloader-specific logic inside the task class, but I don't see `seed_examples` as part of dataloader kwarg in your suggestion..

File: fms_dgt/dataloaders/file.py (hunk 5 above)
Comment (sivasankalpp) [reply fms_dgt/dataloaders/file.py:24]: ah good catch

File: fms_dgt/base/task.py (hunk 4 above)
Comment (mvcrouse) [reply fms_dgt/base/task.py:100]: More what I'm thinking is like how we do it for datastores [here](https://github.com/foundation-model-stack/fms-dgt/blob/970a211939f25d3b9f3c533b84a5860c8be30bb2/fms_dgt/base/task.py#L74). Just have a dict 

`dl_kwargs = { 'seed_examples' : seed_examples, ... }`

And that's passed in like the datastore args like

`**{**dl_kwargs, **dataloader}`

Then it's applied no matter the dataloader

Comment (mvcrouse) [reply fms_dgt/base/task.py:100]: In fact, should we also separate out the initialization of datastores and dataloaders (each individually), just putting those functions into internal methods, so people can overwrite as needed?

Comment (sivasankalpp) [reply fms_dgt/base/task.py]: Ok I updated it! And yeah, we could do that. But it's going to take a bit more refactoring as we need to separate out the kwargs needed for initializing datastores and dataloaders so maybe do that in a separate PR?

File: fms_dgt/base/task.py [hunk 6]
```diff
@@ -87,14 +88,17 @@ def __init__(
...
-            assert (
-                TYPE_KEY in dataloader
-            ), f"Must specify data loader type with '{TYPE_KEY}' key"
-            self._dataloader = get_dataloader(dataloader.pop(TYPE_KEY))(**dataloader)
+            assert TYPE_KEY in dataloader, (
+                "Must specify dataloader type with %s key",
+                TYPE_KEY,
+            )
+            self._dataloader = get_dataloader(dataloader.pop(TYPE_KEY))(
+                **{**dataloader, **dl_kwargs}
```
Comment (mvcrouse) [inline fms_dgt/base/task.py:100]: I'd flip these, that way dl_kwargs doesn't overwrite values from dataloader if they share the same keys, e.g., 

`{ **{3 : 4}, **{3: 5}}` -> {3: 5} whereas we'd want {3 : 4}

Comment (sivasankalpp) [reply fms_dgt/base/task.py]: fixed!

**Review by mvcrouse — APPROVED** [review]
Looks good!

**Pull request merged**